execution-plane/arcon/arcon_macros/src/lib.rs

```diff
@@ -1,27 +1,51 @@
+//! The arcon_macros crate contains macros used by [arcon].
+
 #![recursion_limit = "128"]
 extern crate proc_macro;
 extern crate syn;
 #[macro_use]
 extern crate quote;
 
 use proc_macro::TokenStream;
+use proc_macro2::{Ident, Span};
 use syn::{parse_macro_input, DeriveInput};
 
+/// arcon is a macro that helps define a Arcon supported struct
+///
+/// By default, the macro will implement [std::hash::Hasher] that hashes on all fields of the
+/// struct. Use [arcon_keyed] instead if the struct needs to be hashed on specific fields.
 #[proc_macro_attribute]
 pub fn arcon(metadata: TokenStream, input: TokenStream) -> TokenStream {
     let item = parse_macro_input!(input as DeriveInput);
     let name = &item.ident;
     let _ = proc_macro2::TokenStream::from(metadata);
 
-    if let syn::Data::Struct(_) = item.data {
+    if let syn::Data::Struct(ref s) = item.data {
         let generics = &item.generics;
         let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
 
+        let mut field_hashes = Vec::new();
+        for field in s.fields.iter() {
+            match field.ident {
+                Some(ref ident) => {
+                    let field_hash = quote! { self.#ident.hash(state); };
+                    field_hashes.push(field_hash);
+                }
+                None => panic!("Struct missing identiy"),
+            }
+        }
+
         let output: proc_macro2::TokenStream = {
             quote! {
                 #[derive(Clone, Abomonation)]
                 #item
                 impl #impl_generics ArconType for #name #ty_generics #where_clause {}
+                impl ::std::hash::Hash for #name {
```

Comment:
Yeah, we could do that. I think I forced Hash upon the ArconType to make it work with ChannelStrategy enum. As the KeyBy strategy requires it to have Hash.

Btw, why would it make it unusable to use in a hashmap?

Reply:
I actually don't know if it would become unusable for real, but I'm worried that [this](https://doc.rust-lang.org/stable/std/hash/trait.Hash.html#hash-and-eq) may cause some problem. There may not actually be a huge problem with hashmap, because the user wouldn't probably use that type as a key, but imagine wanting to store values of that type in a hashset. Narrowing the hash impl to use only one field may introduce unnecessary collisions there.

Reply:
I'll open an issue about a custom trait for extracting keys. But merging this for now.